src/sparseml/pytorch/sparsification/quantization/quantize.py

```diff
@@ -214,6 +221,73 @@ def add_input_activation_quant_wrappers(module: Module) -> Module:
     return module
 
 
+def add_output_activation_observers(module: Module):
+    """
+    implementation of torch.quantization add_observers_ that only adds observers
+    according to attached quantization_scheme properties. the existing implementation
+    (1.9+) includes its own logic for propagating including overriding set qconfigs
+    for certain activations without the ability to disable this behavior
+
+    :param module: module to add output activation observers to
+    """
+    # adapted from torch/ao/quantization/quantize.py::_add_observer_
+    # source: https://github.com/pytorch/pytorch/blob/v1.13.0/torch/ao/quantization/quantize.py#L135  # noqa: E501
+    device = next(module.parameters()).device
+
+    def _needs_observer(target_module: Module):
+        if (
+            not hasattr(target_module, "quantization_scheme")
+            or (hasattr(target_module, "activation_post_process"))
+            or isinstance(target_module, torch_quantization.QuantWrapper)
+        ):
+            # submodule not targeted for quantization, already has attached
+            # output observer, or is QuantWrapper (quant wrapper delegates to children)
+            return False
+
+        for descendent_module in target_module.modules():
+            if descendent_module is target_module:
+                continue  # skip itself
+            descendent_scheme = getattr(descendent_module, "quantization_scheme", None)
+            if descendent_scheme is not None and (
+                descendent_scheme.output_activations is not None
+            ):
+                # a descendent of this module targets output activations, return False
+                return False
```

Comment:
It seems like the 1.9 version has the potential to recurse down more than 1 level, whereas this seems to not recurse in this loop. Is the recursion deeper somewhere else or do we not actually need that?

Reply:
they use `children()` which goes 1 layer deep so need to recurse (because they do module swapping through the overrides dict which we don't need). `modules()` goes through all descendents

Reply:
Ohhhhh I did not realize modules was everything! nice simplification